Supervised text mining (#8)
greenelab/knowledge-graph-review: content/03.text-mining.md

1. Mention Clustering Analysis
2. Mention Co-Occurrence approaches

#### Supervised Relationship Extraction

> Supervised extraction uses labeled data to learn patterns and map example input to desired output.

Reviewer: ```suggestion
Supervised extraction uses labeled relationships to learn text patterns that correspond to positively labeled relationships instead of negative ones.
```

> Most of these approaches have flourished due to publically available datasets (Table {@tbl:supervised-text-datasets}).

Reviewer: Labeled ones? I'm guessing that's the important bit, as opposed to just data in general.

> These datasets were constructed by curators for shared open tasks [@raw:biocreative/chemprot; @doi:10.1093/database/baw068] or as a means to provide the scientific community with a gold standard.

Reviewer: I'm guessing there should be one or more references for the second set.

Approaches that use these available datasets range from using support vector machines (SVMs) with custom kernels to deep learning with algorithms that can construct their own features.

> In the rest of this section we discuss approaches that use supervised methods to detect relationship asserting sentences.

Reviewer: I think you might want "...detect sentences asserting a relationship" or "...detect relationship-asserting sentences"

> Extracting relationships in a supervised setting requires the capacity to map textual input onto a high dimensional space.

Reviewer: If you're going to say this, there must be no examples of supervised ML methods that succeed without doing this. Are you sure that's the case? I'm imagining that there might be _some_ features that would be independently predictive (say distance between words).

> Support vector machines are a type of classifier that can accomplish this task by utilizing a mapping function called a kernel [@doi:10.1186/s13326-017-0168-3; @doi:10.1371/journal.pcbi.1004630].

Reviewer: ```suggestion
Support vector machines are a type of classifier that can accomplish this task with a mapping function called a kernel [@doi:10.1186/s13326-017-0168-3; @doi:10.1371/journal.pcbi.1004630].
```

> These kernels take information such as a sentence's dependency tree [@doi:10.1093/database/bay108; @doi:10.1093/bioinformatics/btw503], part of speech tags [@doi:10.1186/s13326-017-0168-3] or even word counts [@doi:10.1371/journal.pcbi.1004630] and directly maps them onto a dense feature space.

Reviewer: ```suggestion
These kernels take information such as a sentence's dependency tree [@doi:10.1093/database/bay108; @doi:10.1093/bioinformatics/btw503], part of speech tags [@doi:10.1186/s13326-017-0168-3] or even word counts [@doi:10.1371/journal.pcbi.1004630] and map them onto a dense feature space.
```

> Within this space a hyperplane can be constructed to separate sentences into the positive class (mentions a relationship) or negative class (does not mention a relationship). 

Reviewer: ```suggestion
Within this space, the methods learn a hyperplane that separates sentences in the positive class (mentions a relationship) from the negative class (does not mention a relationship). 
```

Constructing these kernels necessitates the use of expert knowledge to be successful.

> Similar to regular manual curation, relying on experts to manually construct functions result in highly precise performance, but suffer overall in terms of recall.

Reviewer: I don't quite understand this - do you mean that the kernels are manually constructed/selected? And does it really give performance similar to manual curation? I'd expect them to be pretty different.

Besides single kernel methods, a recent study used an ensemble of SVMs to extract disease-gene associations [@doi:10.1371/journal.pone.0200699].

> This provided powerful preformance in terms of precision and outperformed the various baseline models.

Reviewer: typo in performance, but more importantly what do you mean by `powerful`?

> Besides single kernel methods, a recent study used an ensemble of SVMs to extract disease-gene associations [@doi:10.1371/journal.pone.0200699].

Reviewer: ```suggestion
In addition to single kernel methods, a recent study used an ensemble of SVMs to extract disease-gene associations [@doi:10.1371/journal.pone.0200699].
```

> This ensemble outperformed notable disease-gene association extractors [@doi:10.1016/j.jbi.2015.08.008; @doi:10.1186/s12859-015-0472-9] in terms of precision, recall and even F1 score.

Reviewer: ```suggestion
The ensemble outperformed notable disease-gene association extractors [@doi:10.1016/j.jbi.2015.08.008; @doi:10.1186/s12859-015-0472-9] in terms of precision, recall and F1 score.
```

Overall, future SVM methods should consider a more automated approach to construct these kernels or develop ways to use semi-automatic methods for kenel construction.

> Deep learning is an increasingly popular field that can construct their own features within a high dimensional space [@raw:GoodfellowDL; @doi:10.1038/nature14539].

Reviewer: "field" -> maybe "approach" or "class of techniques" or something :)

> Overall, future SVM methods should consider a more automated approaches to define these kernel functions or develop ways to use semi-automatic methods for kernel construction.

Reviewer: Are there people doing stuff like this? I feel like deep NNs kind of accomplish the same thing (heavily nonlinear mapping of the feature space). Perhaps it would be better to just note that deep NNs have kinda taken over, which takes you to your next paragraph?

> Deep learning is an increasingly popular class of techniques that can construct their own features within a high dimensional space [@raw:GoodfellowDL; @doi:10.1038/nature14539].

Reviewer: This would be a good transition in if you modify the final sentence of the previous paragraph to note that being able to do nonlinear feature combination automatically is valuable.

> These methods use different forms of neural networks, such a recurrent neural network or a convolutional neural network, to perform classification.

Reviewer: ```suggestion
These methods use different forms of neural networks, such as recurrent or convolutional neural networks, to perform classification.

```

Recurrent neural networks (RNN) are designed for sequential analysis that consist of using a repeatedly updating hidden state to make predictions.
An example of a recurrent neural network is a long short term memory (LSTM) network [@doi:10.1162/neco.1997.9.8.1735].
Cocos et al [@doi:10.1093/jamia/ocw180] used a LSTM to extract drug side effects from de-identified twitter posts, while Yadav et al [] used an LSTM to extract protein protein interactions.
Other works have used LSTMs to perform relationship extraction [@arxiv:1708.03743; @doi:10.1093/bioinformatics/btw486; @arxiv:1808.09101; @doi:10.1186/s12859-017-1609-9; @doi:10.1093/jamia/ocw180]. 
Despite the success of these networks, runtime for RNNs can be an issue.

> Using a LSTM on a dataset that has 200+ token sentences can take up to days to run.

Reviewer: reference

> Based on this future approaches should be cautious on length of each individual sentence using this type of network.

Reviewer: I think you mean people using these approaches should be cautious - not the approaches themselves. Are there any refs for good ways to deal with this? Drop long sentences? Something else?

> Furthermore, these methods require a lot of data to perform well.

Reviewer: Refs? How much?

> Future methods should consider solutions to obtaining these large quantities of data.

Reviewer: Are these the type of things you discuss below with weak/distant supervision?

Reviewer: Also: do transfer learning approaches have a role here?

Convolutional neural networks (CNN) are heavily used in image analysis and consist of using multiple kernel filters to capture small subsets of the overall image [@doi:10.1038/nature14539].

> In the context of text mining an image is repleaced with words within a sentence mapped to dense vectors (i.e. word embeddings) [@arxiv:1301.3781; @arxiv:1310.4546].

Reviewer: `repleaced` -> `replaced`

Peng et al. [@arxiv:1706.01556v2] used a CNN to extract sentences concerning protein-protein intercations and Zhou et al. [@doi:10.1186/s12859-019-2873-7] used a CNN to extract chemical-disease relations.

> Other work have used this network for extraction [@doi:10.1177/0165551516673485; @doi:10.1093/database/bay073; @doi:10.1101/730085]

Reviewer: `work` -> researchers? practitioners? something like that since it's not the work doing the reuse but people.

> Just like RNNs, these networks perform well when a millions amount of data is present [@arxiv:1707.02968]. 

Reviewer: ```suggestion
Just like RNNs, these networks perform well when millions of labeled examples are present [@arxiv:1707.02968]. 
```

Just like RNNs, these networks require a large amount of data to peform well.

A solution for the necessity of large datasets is to use semi-supervised or weakly supervised methods. 
Semi-supervised learning combines labeled data with unlabeled data to extract relationships.
For example, one study used a variational auto encoder with a LSTM network to extract protein-protein interactions from pubmed abstracts and full text [@arxiv:1901.06103v1].
This is an elogant solution to handle the small dataset problem, but requires to have labeled data to start. 
The dependency on labeled data can make finding under-studied relationships diffcult as one would need to find examples of these missing relationships at the start.

> 

Reviewer: This seems like it belongs in the same paragraph with the above, at least as the topic sentence for the previous paragraph is written.

> Future approaches that use CNNs or Recurrent neural networks should consider solutions to obtaining these large quantities of data through means such as weak supervision [@doi:10.3115/1690219.1690287], semi-supervised learning [@doi:10.2200/S00196ED1V01Y200906AIM006] or using pre-trained networks via transfer learning [@doi:10.1109/TKDE.2009.191; @doi:10.1186/s40537-016-0043-6].

Reviewer: ```suggestion
Future approaches that use CNNs or RNNs should consider solutions to obtaining these large quantities of data through means such as weak supervision [@doi:10.3115/1690219.1690287], semi-supervised learning [@doi:10.2200/S00196ED1V01Y200906AIM006] or using pre-trained networks via transfer learning [@doi:10.1109/TKDE.2009.191; @doi:10.1186/s40537-016-0043-6].
```

If you start using an acronym, buy into it entirely so your readers learn it :)

Weak or distant supervision is an approach that uses noisy or even erroneous labels to perform supervised relation extraction [@doi:10.3115/1690219.1690287].
Algorithms under this paradigm can learn through this noise and produce comparable performance.

> Notable works related to this paradigm can be found here: 

Reviewer: I think you will need to briefly summarize these contributions + strengths/weaknesses. It'd also be good to discuss what these approaches need for success.

> Semi-supervised learning [@doi:10.2200/S00196ED1V01Y200906AIM006] and weak supervision [@doi:10.3115/1690219.1690287] are techniques that can construct large datasets for machine learning classifiers to use. 

Reviewer: ```suggestion
Semi-supervised learning [@doi:10.2200/S00196ED1V01Y200906AIM006] and weak supervision [@doi:10.3115/1690219.1690287] are techniques that can construct large datasets for machine learning classifiers. 
```

> Semi-supervised learning consist of combining labeled data with unlabeled data to extract relationships.

Reviewer: ```suggestion
Semi-supervised learning consists of combining labeled data with unlabeled data to extract relationships.
```

[@doi:10.1093/bioinformatics/btv476; @doi:10.14778/3157794.3157797; @doi:10.1145/3209889.3209898; @doi:10.1101/730085].
These approaches provide promising results in terms of relation extraction.
Future methods should consider using more of this paradigm as well as the semi-supervised paradigm to perform extraction.

| Dataset | Type of Sentences |
| --- | --- |
| [@doi:10.1016/j.artmed.2004.07.016] | PPI |
| [@doi:10.1186/1471-2105-8-50] | PPI | 
| [@raw:LLL] | PPI |
| [@raw:IEPA] | PPI |
| [@doi:10.1093/bioinformatics/btl616] | PPI |
| [@doi:10.1016/j.jbi.2012.04.004] | DaG |
| [@doi:10.1186/s12859-015-0472-9] | DaG |
| [@doi:10.1186/1471-2105-14-323] | DaG | 
| [@doi:10.1186/1471-2105-13-161] | DaG |
| [@doi:10.1093/database/baw068] | Compound induces Disease |
| [@raw:biocreative/chemprot] | CbG |

> Table: A set of publicaly available datasets for supervised text mining. {#tbl:supervised-text-datasets}

Reviewer: `publicaly` spelling

> Some approaches have attempted to fix this noise problem [@arxiv:1805.09927; @arxiv:1805.09929; @doi:10.18653/v1/W17-2323], but it still remains an open problem to be solved.

Reviewer: Do you want to cite your own paper on label function reuse?

> | [@doi:10.1016/j.artmed.2004.07.016] | PPI |

Reviewer: I think you're going to want to provide names for these - not just numerical references.

Recurrent neural networks (RNN) are designed for sequential analysis that consist of using a repeatedly updating hidden state to make predictions.
An example of a recurrent neural network is a long short term memory (LSTM) network [@doi:10.1162/neco.1997.9.8.1735].
Cocos et al [@doi:10.1093/jamia/ocw180] used a LSTM to extract drug side effects from de-identified twitter posts, while Yadav et al. [@doi:10.3233/978-1-61499-830-3-644] used an LSTM to extract protein-protein interactions.
Other works have used LSTMs to perform relationship extraction [@arxiv:1708.03743; @doi:10.1093/bioinformatics/btw486; @arxiv:1808.09101; @doi:10.1186/s12859-017-1609-9; @doi:10.1093/jamia/ocw180]. 
Despite the success of these networks, training can be difficult as these networks are highly susceptible to vanishing and exploding gradients [@doi:10.1109/ICNN.1993.298725; @arxiv:1808.03314].
One solution to this problem is to clip the gradients while the neural network trains [@arxiv:1211.5063].
Besides the gradient problem, these approaches peak in performance when the dataset reaches at least a tens of thousand of data points [@arxiv:1707.02968].

> Convolutional neural networks (CNN) are heavily used in image analysis and consist of using multiple kernel filters to capture small subsets of the overall image [@doi:10.1038/nature14539].

Reviewer: ```suggestion
Convolutional neural networks (CNNs), which are widely applied for image analysis, use multiple kernel filters to capture small subsets of an overall image [@doi:10.1038/nature14539].
```

This is an elegant solution to handle the small dataset problem, but requires labeled data to start. 
The dependency on labeled data makes finding under-studied relationships difficult as one would need to find or construct examples of the missing relationships in the beginning.

Weak or distant supervision takes a different approach that uses noisy or even erroneous labels to train classifiers [@doi:10.3115/1690219.1690287; @doi:10.1093/bioinformatics/btv476; @doi:10.14778/3157794.3157797; @doi:10.1145/3209889.3209898].
Under this paradigm sentences are labeled based on their mention pair being present (positive) or absent (negative) in a database.
Once these labels are obtained a machine learning classifier can now be trained to predict sentences [@doi:10.3115/1690219.1690287].
For example, Thomas et al. [@raw:Thomas2011LearningTE] used distant supervision to train a support vector machine to extract sentences mentioning protein-protein interactions (ppi). 
Their SVM model achieved comparable performance against a baseline model; however, the noise generated via distant supervision was difficult to eradicate [@raw:Thomas2011LearningTE].

> Some approaches have attempted to fix this noise problem [@arxiv:1805.09927; @arxiv:1805.09929; @doi:10.18653/v1/W17-2323; @doi:10.1101/730085], but it still remains an open problem to be solved.

Reviewer: ```suggestion
A number of efforts have focused on combining distant supervision with other types of labeling strategies to reduce the negative impacts of noisy knowledge bases [@arxiv:1805.09927; @arxiv:1805.09929; @doi:10.18653/v1/W17-2323].
Nicholson et al. [@doi:10.1101/730085] found that, in some circumstances, these strategies and rules can be reused across different types of biomedical edges to learn a heterogeneous knowledge graph if those edges describe similar physical concepts.
This remains an active area of investigation with numerous associated challenges and opportunities.
```
